mpas_analysis/ocean/meridional_overturning_circulation.py

```diff
@@ -522,41 +502,40 @@ def write_file(dsMOCTimeSeries):
 #     return (mocDictClimo, mocDictTseries)
 
 
-def _compute_transport(maxEdgesInTransect, transectEdgeGlobalIDs,
-                       transectEdgeMaskSigns, nz, dvEdge, refLayerThickness,
+def _compute_transport(transectEdgeMaskSigns, nz, dvEdge, refLayerThickness,
                        horizontalVel):  # {{{
 
     '''compute mass transport across southern transect of ocean basin'''
 
-    transportZEdge = np.zeros([nz, maxEdgesInTransect])
-    for i in range(maxEdgesInTransect):
-        if transectEdgeGlobalIDs[i] == 0:
-            break
-        # subtract 1 because of python 0-indexing
-        iEdge = transectEdgeGlobalIDs[i] - 1
-        transportZEdge[:, i] = horizontalVel[iEdge, :] * \
-            transectEdgeMaskSigns[iEdge, np.newaxis] * \
-            dvEdge[iEdge, np.newaxis] * \
-            refLayerThickness[np.newaxis, :]
-    transportZ = transportZEdge.sum(axis=1)
+    transect = transectEdgeMaskSigns != 0
+    transportZ = (refLayerThickness*horizontalVel.where(transect)
+                  *transectEdgeMaskSigns.where(transect)
+                  *dvEdge.where(transect)
+                  ).sum('nEdges')
+
     return transportZ  # }}}
 
 
 def _compute_moc(latBins, nz, latCell, regionCellMask, transportZ,
                  velArea):  # {{{
 
     '''compute meridionally integrated MOC streamfunction'''
-
-    mocTop = np.zeros([np.size(latBins), nz+1])
-    mocTop[0, range(1, nz+1)] = transportZ.cumsum()
-    for iLat in range(1, np.size(latBins)):
-        indlat = np.logical_and(np.logical_and(
-                     regionCellMask == 1, latCell >= latBins[iLat-1]),
-                     latCell < latBins[iLat])
-        mocTop[iLat, :] = mocTop[iLat-1, :] + velArea[indlat, :].sum(axis=0)
+    # note computation should be more cleanly facilitated 
+    # via a multidimensional groupby_bins when available
+    # (see https://github.com/pydata/xarray/pull/924)
+
+    mocTop = xr.DataArray(np.zeros([len(nz), len(latBins)]),
+                          coords=[nz, latBins])
+    mocTop[1:, 0] = transportZ.cumsum(axis=0)
+    for iLat in np.arange(1, len(latBins)):
+        mask = np.logical_and(
+                np.logical_and(latCell >= latBins[iLat-1],
+                               latCell < latBins[iLat]),
+                               regionCellMask > 0)
+        mocTop[:, iLat] = velArea.where(mask).sum('nCells')
+    mocTop = mocTop.cumsum('latBins')
```

Comment:
So just for the record, I did some profiling of this code as part of my work on computing and caching the MOC for each month.  This function is, indeed, by far the slowest part of the code besides loading the data files in the first place.  So here is were we might focus our efforts on optimizing in the future.

By the way, the results of computing and caching monthly MOC were *so, so very* slow that I will be deleting that branch and we will never speak of it again...